### PR TITLE
QUIC: Limit the sending after 3 packets with unavlidated path

### PR DESCRIPTION
**Pull request by scw00**
part of #4905

### CODE REVIEW

File: iocore/net/QUICNetVConnection.cc [hunk 1]
```diff
@@ -1226,21 +1226,30 @@ QUICNetVConnection::_state_common_send_packet()
         continue;
       }
 
+      uint32_t max_packet_size = udp_payload_len - written;
       if (this->netvc_context == NET_VCONNECTION_IN && !this->_is_src_addr_verified() &&
           this->_handshake_packets_sent >= MAX_PACKETS_WITHOUT_SRC_ADDR_VARIDATION) {
```
Comment (masaori335) [inline iocore/net/QUICNetVConnection.cc]: Do we still need this condition?

**Comment by scw00** [comment]
Hi @masaori335  Updated. Separating the `AddrState` into independent file because of test case needed.

File: iocore/net/quic/QUICAddrVerifyState.cc [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  }
+
+  uint64_t tmp = this->_windows;
+  if (tmp + windows > UINT32_MAX) {
+    // overflow
+    this->_windows = UINT32_MAX;
+    return;
+  }
+
+  this->_windows += windows;
```
[the remark this answers is not in the document]
Comment (masaori335) [reply iocore/net/quic/QUICAddrVerifyState.cc]: FWIW, below is what we haven't followed change at draft-13. ( words were clarified a lot, though )

```
- Servers MUST NOT send more than three Handshake **packets** without receiving a packet from a verified source address.
+ Servers MUST NOT send more than three **datagrams** including Initial and Handshake packets without receiving a packet from a verified source address.
```

https://tools.ietf.org/rfcdiff?url2=draft-ietf-quic-transport-13.txt

**Comment by scw00** [comment]
done !

File: iocore/net/quic/QUICAddrVerifyState.h [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+ */
+
+#pragma once
+
+#include <cstdint>
+
+class QUICAddrVerifyState
+{
+public:
+  QUICAddrVerifyState(uint8_t packets, uint32_t windows) : _windows(windows) {}
```
Comment (masaori335) [inline iocore/net/quic/QUICAddrVerifyState.h]: [nitpicking] This constructor looks not used.

File: iocore/net/QUICNetVConnection.cc [hunk 4]
```diff
@@ -1052,8 +1050,8 @@ QUICNetVConnection::_state_handshake_process_handshake_packet(QUICPacketUPtr pac
 {
   // Source address is verified by receiving any message from the client encrypted using the
```
Comment (maskit) [inline iocore/net/QUICNetVConnection.cc]: Receiving a message encrypted using Handshake key is not the only way. Address validation can be done with address validation tokens. I'm not going to stop landing this, but it should be fixed in the future.

I haven't thought this deeply, but I would make something like these:
- QUICAddressValidator
- QUICPacketSizeLimiter interface
- QUICCongestionController extends QUICPacketSizeLimiter
- QUICAddressValidator extends QUICPacketSizeLimiter
- QUICPathValidator extends QUICPacketSizeLimiter
- QUICPMTUDiscoverer extends QUICPacketSizeLimiter

Comment (masaori335) [reply iocore/net/QUICNetVConnection.cc]: Right. Please file an issue.

Comment (maskit) [reply iocore/net/QUICNetVConnection.cc]: Filed one for address validation with tokens. #4930

**Pull request merged**